bdns/dns.go

```diff
@@ -488,37 +497,54 @@ func isPrivateV6(ip net.IP) bool {
 	return false
 }
 
-func (dnsClient *impl) lookupIP(ctx context.Context, hostname string, ipType uint16) ([]dns.RR, error) {
-	resp, err := dnsClient.exchangeOne(ctx, hostname, ipType)
+func (dnsClient *impl) lookupIP(ctx context.Context, hostname string, ipType uint16) ([]dns.RR, ResolverAddr, error) {
+	resp, resolver, err := dnsClient.exchangeOne(ctx, hostname, ipType)
 	errWrap := wrapErr(ipType, hostname, resp, err)
 	if errWrap != nil {
-		return nil, errWrap
+		return nil, resolver, errWrap
 	}
-	return resp.Answer, nil
+	return resp.Answer, resolver, nil
 }
 
 // LookupHost sends a DNS query to find all A and AAAA records associated with
 // the provided hostname. This method assumes that the external resolver will
 // chase CNAME/DNAME aliases and return relevant records. It will retry
 // requests in the case of temporary network errors. It returns an error if
 // both the A and AAAA lookups fail or are empty, but succeeds otherwise.
-func (dnsClient *impl) LookupHost(ctx context.Context, hostname string) ([]net.IP, error) {
+func (dnsClient *impl) LookupHost(ctx context.Context, hostname string) ([]net.IP, ResolverAddr, error) {
 	var recordsA, recordsAAAA []dns.RR
 	var errA, errAAAA error
+	var resolverA, resolverAAAA ResolverAddr
 	var wg sync.WaitGroup
 
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		recordsA, errA = dnsClient.lookupIP(ctx, hostname, dns.TypeA)
+		recordsA, resolverA, errA = dnsClient.lookupIP(ctx, hostname, dns.TypeA)
+		if resolverA != "" {
+			resolverA = "A:" + resolverA
+		}
 	}()
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		recordsAAAA, errAAAA = dnsClient.lookupIP(ctx, hostname, dns.TypeAAAA)
+		recordsAAAA, resolverAAAA, errAAAA = dnsClient.lookupIP(ctx, hostname, dns.TypeAAAA)
+		if resolverAAAA != "" {
+			resolverAAAA = "AAAA:" + resolverAAAA
+		}
 	}()
 	wg.Wait()
 
+	var resolvers []string
+	if resolverA.String() != "" {
+		resolvers = append(resolvers, resolverA.String())
+	}
+	if resolverAAAA.String() != "" {
+		resolvers = append(resolvers, resolverAAAA.String())
+	}
```

Comment:
I believe that you can just treat these like any other `string`s.

```suggestion
	var resolvers []string
	if resolverA != "" {
		resolvers = append(resolvers, resolverA)
	}
	if resolverAAAA != "" {
		resolvers = append(resolvers, resolverAAAA)
	}
```

Reply:
You cannot.

This portion works with a type cast, which is fine. I removed the stringer.
```
	var resolvers []string
	if resolverA != "" {
		resolvers = append(resolvers, string(resolverA))
	}
```

This part throws an error.
```
	if resolverAAAA != "" {
		resolvers = append(resolvers, resolverAAAA)
	}


cannot use resolverAAAA (variable of type ResolverAddr) as string value in argument to appendcompiler[IncompatibleAssign](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#IncompatibleAssign)
```

Reply:
https://github.com/letsencrypt/boulder/pull/7271#discussion_r1474759109